csrc/id_model/id_model.cpp

```diff
@@ -948,4 +956,265 @@ ValGraph IdModel::buildIntersection(
+    auto tmp_inputs = new_inputs;
+    for (const auto i : c10::irange(new_inputs.size())) {
+      new_inputs.at(i) = IterDomainBuilder(tmp_inputs.at(i))
+                             .iter_type(IterType::Iteration)
+                             .build();
+      id_definitions_[new_inputs.at(i)];
+      id_uses_[new_inputs.at(i)];
+      for (auto mode : initialized_modes) {
+        idGraph(mode).initializeVal(new_inputs.at(i), {}, {});
+        idGraph(mode).mapVals(new_inputs.at(i), tmp_inputs.at(i));
+      }
+    }
+  }
+
+  const std::vector<IterDomain*> orig_input_ids =
+      ir_utils::filterByType<IterDomain>(expr->inputs()).vector();
+
+  // Sanity check of the original inputs
+  {
+    NVF_ERROR(
+        new_inputs.size() == orig_input_ids.size(),
+        "Invalid number of inputs: ",
+        new_inputs.size(),
+        " does not match number of iter domain inputs for ",
+        expr->toString());
+
+    for (auto mode : initialized_modes) {
+      for (auto inp : orig_input_ids) {
+        NVF_ERROR(
+            idGraph(mode).hasGroup(inp),
+            "All inputs for replay need to be initialized in all graphs, ",
+            inp->toString(),
+            " was not found in mode: ",
+            mode);
+      }
+    }
+  }
+
+  // Create the new expression with provided inputs
+  auto replay = ReplayTransform::replayAs(new_inputs, expr);
+
+  for (auto out_id : ir_utils::filterByType<IterDomain>(replay->outputs())) {
+    id_definitions_[out_id].pushBack(replay);
+    id_uses_[out_id];
+  }
+
+  // Add the expression to the uses of the inputs
+  for (auto inp_id : ir_utils::filterByType<IterDomain>(replay->inputs())) {
+    id_definitions_[inp_id];
+    id_uses_[inp_id].pushBack(replay);
+  }
+
+  // Initialize output iter domains in the graphs
+  for (auto mode : initialized_modes) {
+    idGraph(mode).registerExpr(replay);
+    auto replay_group = idGraph(mode).toGroup(replay);
+
+    // Initialize output ids in map
+    for (auto out_id : ir_utils::filterByType<IterDomain>(replay->outputs())) {
+      idGraph(mode).initializeVal(out_id, {replay}, {});
+    }
+
+    // Update uses of the inputs in the graphs
+    for (auto inp_id : ir_utils::filterByType<IterDomain>(replay->inputs())) {
+      auto inp_group = idGraph(mode).toGroup(inp_id);
+      idGraph(mode).addUniqueUses(inp_group, replay_group);
+    }
+
+    // Propagate through all the uses of the iter domain groups of the inputs
+    // with the new expression.
+    auto& graph = idGraph(mode);
+    // Gather all use expressions from inputs
+    VectorOfUniqueEntries<Expr*> representative_uses;
+    for (IterDomain* inp : new_inputs) {
+      for (const ExprGroup& use_group : graph.getUses(graph.toGroup(inp))) {
+        NVF_ERROR(!use_group->empty());
+        representative_uses.pushBack(use_group->front());
+      }
+    }
+
+    for (auto rep_use : representative_uses) {
+      graph.maybeMapThroughExprs(rep_use, replay, true);
+    }
```

Comment:
I am trying to understand why we need a `maybeMapThroughExprs` here. Is this `maybeMapThroughExprs` called here for the purpose of adding the replayed outputs to the loop and permissive disjoint set that it should belong to?